test: add warehouse application

davidkostov opened the pull request:
Added a warehouse application, the warehouse can store, sell and
list any and all contents of its inventory.

mgenov commented on src/com/clouway/testing/Warehouse.java:
```diff
@@ -0,0 +1,103 @@
+    }
+
+    private boolean typeIsFull(Stock theStock) {
+        switch (theStock.getType()) {
```
This code smells as when you add a new stock you have to modify this method. This also breaks the Open Closed Principle that we discussed.

mgenov commented on src/com/clouway/testing/Stock.java:
```diff
@@ -0,0 +1,65 @@
+        return price;
+    }
+
+    public Double getWeight() {
```
It's more readable negate expression to be first. 

```java

if (weight == null) { //e.g code reads: when weight is null, then ISE is thrown 
  throw new IllegalStateException("The weight for this product has not been provided!");
}
return weight;
```

mgenov commented on src/com/clouway/testing/Warehouse.java:
```diff
@@ -0,0 +1,103 @@
+ */
+public class Warehouse {
+
+    private final Integer maxFood = 3000, maxClothing = 1000,
```
If I have to add a new product I have to add a new constant here which smells. 

The stock inventory could be passed as params to the warehouse.

mgenov commented on src/com/clouway/testing/Warehouse.java:
```diff
@@ -0,0 +1,103 @@
+                current.setQuantity(current.getQuantity() - quantity);
+
+                switch (current.getPriceUnit()) {
+                    case "kg":
```
The same issue here with the measurement units.

mgenov commented on src/com/clouway/testing/Warehouse.java:
```diff
@@ -0,0 +1,103 @@
+        throw new IllegalArgumentException(name + " - not present in the warehouse`s inventory!");
+    }
+
+    public void sortByPrice() {
```
Warehouse.sortByPrice -> sort what by price ?, e.g name is misleading

mgenov commented on test/com/clouway/testing/WarehouseTest.java:
```diff
@@ -0,0 +1,64 @@
+    }
+
+    @Before
+    public void setUpStream() {
```
Move setup/teardown methods before tests.

mgenov commented on test/com/clouway/testing/WarehouseTest.java:
```diff
@@ -0,0 +1,64 @@
+    private PrintStream stdout = System.out;
+
+    @Test
+    public void sellStockWorks() {
```
You are not testing when limit is reached.

davidkostov commented:
Fixed. Dropped the idea of calculating profit because depending on the unit there are too many possible calculations to be made. Plus I am not sure what limit you were talking about so I tested receiving stock over limit.

mgenov commented on test/com/clouway/testing/WarehouseTest.java:
```diff
@@ -0,0 +1,76 @@
+
+    @Test
+    public void printProductsSortedByPriceWorks() {
+        warehouse.clearInventory();
```
warehouse should start with clear inventory by default I suppose.

davidkostov replied:
Yes but I am using a local warehouse variable for all tests to keep the code short, so I am clearing every time to start the test clean. But I can make a new warehouse every time if you want. I just think it is more readable this way.

mgenov replied:
I'm not sure that I understand you correctly. 

JUnit creates a new instance of the test before each test run which means a new instance of Warehouse is going to be created as it's defined as package level variable.

mgenov commented on test/com/clouway/testing/WarehouseTest.java:
```diff
@@ -0,0 +1,76 @@
+
+        warehouse.sortInventoryByPrice();
+        warehouse.printProducts();
+        assertThat(outContent.toString(),
```
```java
List<Stock> inventoryStocks = 
asserThat(warehouse.geStocksSortedByPrice(), is(equalTo(Arrays.asList(new Stock(...), new Stock(...)));
```

mgenov commented on src/com/clouway/testing/Warehouse.java:
```diff
@@ -0,0 +1,76 @@
+                    type.getQuantity() + newStock.getQuantity() < type.getMaxQuantity()) {
+
+                if (type.getName().equals(newStock.getType())) {
+                    type.setQuantity(type.getQuantity() + newStock.getQuantity());
```
mutation is not a good thing. It's better to create a new instance with the new value.

davidkostov commented:
Fixed, but removing the mutation resulted in lowering readability a lot.

mgenov commented on src/com/clouway/testing/Warehouse.java:
```diff
@@ -0,0 +1,75 @@
+                            return;
+                        }
+                    }
+                    stockTypes.set(i, new StockType(typeName,typeQuantity+newStock.getQuantity()));
```
```java
stockTypes.set(stockType.changeQuantityTo(newStock.getQuuantity())) 
```

i.e the method of stockType could be responsible for the creation.

mgenov commented on src/com/clouway/testing/Warehouse.java:
```diff
@@ -0,0 +1,75 @@
+    }
+
+    public void receive(Stock newStock) {
+        for (int i=0; i<stockTypes.size(); i++) {
```
You can extract private method for finding of stock Type: 

StockType targetStock = stockTypes.get(findStockTypeIndexByName(newStock.getName()))

This will clean things up a little bit in this method. Another thing that could be improved is Inventory to be represented as object instead of list. 

inventory.updateQuantity(stockName, newQuantity) 

Primitive obsession to List is forcing your logic to become complicated.

davidkostov commented:
Fixed.

mgenov commented on src/com/clouway/testing/Inventory.java:
```diff
@@ -0,0 +1,70 @@
+        throw new IllegalArgumentException("There is no such stock type in this inventory!");
+    }
+
+    public Integer findStockIndexByName(String stockName){
```
If this method is in the inventory you have to expose stock list to the callers as indexes are returned. This doesn't sound good as the clients has to deal with one more thing. 

The following statement is much simpler: 
```java
inventory.amendStock("apples", 20) 
```

mgenov replied:
.

mgenov commented on src/com/clouway/testing/Warehouse.java:
```diff
@@ -0,0 +1,61 @@
+ */
+public class Warehouse {
+
+    Inventory inventory;
```
private final Inventory inventory;

mgenov commented on src/com/clouway/testing/Warehouse.java:
```diff
@@ -0,0 +1,61 @@
+    }
+
+    public void sell(String name, Integer quantity) {
+        Integer targetStockIndex = inventory.findStockIndexByName(name);
```
The deisgn here is going to change but It's more readable to use negative condition instead 

```java
if (targetStockIndex == null) {
  throw new IllegalStateException("There aren't enough " + targetStock.getName() + " in the stock");
}
..

```

mgenov commented on src/com/clouway/testing/Inventory.java:
```diff
@@ -0,0 +1,70 @@
+        contents.add(newStock);
+    }
+
+    public void amendStockType(Integer index, Integer quantity){
```
Adding of a new stock at runtime is not actually required by the task. If you insist to be like that I don't mind to be kept.

davidkostov commented:
Fixed.

mgenov commented:
I'm not able to track easily your changes, could you write comment responses with info what and how was changed. It would be easier for me to track it.

davidkostov commented:
Done.

mgenov commented on src/com/clouway/testing/Warehouse.java:
```diff
@@ -0,0 +1,70 @@
+
+    public void receive(Stock newStock) {
+        //removed most of the declarations that i don`t need after fixing amend methods
+        StockType targetType = inventory.getStockTypes().get(
```
Just: StockType targetType = inventory.findStockByName(newStock.getType()); 

You don't have to expose internal behaviour of inventory.

davidkostov replied:
Done. But I also needed the find index methods so i kept those.

mgenov replied:
Not sure about that: https://martinfowler.com/bliki/TellDontAsk.html

davidkostov commented:
Fixed.

mgenov commented on src/com/clouway/testing/Inventory.java:
```diff
@@ -0,0 +1,96 @@
+        this.stockTypes = stockTypes;
+    }
+
+    private Integer findStockTypeIndexByName(String typeName) {
```
private methods should go after the public ones. 

From: https://github.com/clouway/teamspace/wiki/Java-Code-Convention

```java
public abstract class Foo {
  // Type declarations.
  public class FooBaz {
  }
 
  private class FooBar {
  }

  // Static field declarations.
  // Remember, fields do NOT need to be sorted.
  static String B;
  static String A;

  // Static initializer declarations.
  static {
  }

  // Static methods declarations.
  // Remember, methods do need to be sorted.
  static void aStatic() {
  }

  static void bStatic() {
  }

  // Instance field declaration.
  String bField;
  String aField;

  // Instance Initializer declarations.
  {
  }

  // Constructors declaration.
  public Foo() {
  }

  protected Foo(String s) {
  }

  Foo(int i) {
  }

  private Foo(boolean b) {
  }

  // Instance method declaration.
  public void b() {
  }

  public void c() {
  }

  protected void a() {
  }

  protected void d() {
  }

  protected void e() {
  }

  protected void f() {
  }

  String h() {
  }

  // The "abstract" keyword does not modify the position of the method.
  abstract String i();

  void j() {
  }

  private void g() {
  }
}
```

mgenov commented on src/com/clouway/testing/Inventory.java:
```diff
@@ -0,0 +1,96 @@
+        stockTypes.set(index, amendedStockType);
+    }
+
+    public List<StockType> getStockTypes() {
```
This method should not be required anymore if amend methods are doing there job and finding is explicit.

davidkostov commented:
Fixed.

mgenov commented on src/com/clouway/testing/Inventory.java:
```diff
@@ -0,0 +1,61 @@
+    }
+
+    public Stock findStockByName(String stockName) {
+        for (int i = 0; i < contents.size(); i++) {
```
```java
for (Stock each : contents) {
  if (each.getName().equals(stockName)) {
       return each
   }
}
return null
```

btw: Optional is a something that we use instead of null
http://www.oracle.com/technetwork/articles/java/java8-optional-2175753.html

mgenov commented on test/com/clouway/testing/WarehouseTest.java:
```diff
@@ -0,0 +1,62 @@
+    }
+
+    @Test
+    public void printProductsSortedByPriceWorks() {
```
comparison by toString() is dangerous and is not something that should be used.

You can compare directly objects if they override equals

davidkostov replied:
After overriding equals for Inventory, Stock, StockType and comparing two objects of Inventory type(they are absolutely the same!), I still get a miss match that is one blank line...
![image](https://user-images.githubusercontent.com/29853955/29312580-c2598a1c-81be-11e7-9dc5-79e99845e644.png)

davidkostov replied:
After overriding equals for Inventory, Stock, StockType and comparing two objects of Inventory type(they are absolutely the same!), I still get a miss match that is one blank line... I will push as it is so you can take a look.
![image](https://user-images.githubusercontent.com/29853955/29312580-c2598a1c-81be-11e7-9dc5-79e99845e644.png)

davidkostov commented:
Managed to fix "printSortedByPrice" test by comparing the lists of stock instead of inventories(as it was initialy intended). Also made use of Optional instead of returning null. Added Checker class that i use to check if the stock/type exists before doing things with it. Override-ed the equals method for Inventory, Stock, StockType classes.
